server/src/main/kotlin/io/spine/server/security/Security.kt

```diff
@@ -0,0 +1,122 @@
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.security
+
+import io.spine.code.java.PackageName
+import io.spine.server.Server
+import io.spine.server.security.CallerProvider.previousCallerClass
+import io.spine.system.server.SystemContext
+import java.lang.StackWalker.Option.RETAIN_CLASS_REFERENCE
+import java.lang.StackWalker.StackFrame
+import java.util.stream.Stream
+
+/**
+ * Controls which class is allowed to call a method.
+ */
+public object Security {
+
+    private const val TESTING_SERVER_PACKAGE = "io.spine.testing.server"
+
+    /**
+     * Throws [SecurityException] of the calling class does not belong to
+     * the Spine Event Engine framework or its testing utilities.
+     */
+    @JvmStatic
+    public fun allowOnlyFrameworkServer() {
+        val callingClass = previousCallerClass()
+        if (!belongsToServer(callingClass)) {
+            throw nonAllowedCaller(callingClass)
+        }
+    }
+
+    /**
+     * Tells if the calling class belongs to the server packages for which a call is permitted.
+     *
+     * The allowed packages are:
+     *  1. [io.spine.server]
+     *  2. [io.spine.system.server]
+     *  3. [io.spine.testing.server][Security.TESTING_SERVER_PACKAGE]
+     *
+     * @param callingClass The class to check.
+     * @return `true` if the class is allowed to make a call, `false` otherwise.
+     */
+    private fun belongsToServer(callingClass: Class<*>): Boolean {
+        val serverPackage = PackageName.of(Server::class.java)
+        val systemServerPackage = PackageName.of(
+            SystemContext::class.java
+        )
+        val callingClassName = callingClass.name
+        val result =
+            callingClassName.startsWith(serverPackage.value())
+                    || callingClassName.startsWith(systemServerPackage.value())
+                    || callingClassName.startsWith(TESTING_SERVER_PACKAGE)
+        return result
+    }
+
+    private fun nonAllowedCaller(callingClass: Class<*>): SecurityException {
```

Comment:
```suggestion
    private fun notAllowedCaller(callingClass: Class<*>): SecurityException {
```

I have not seen `non allowed` before. You can save on braces here.